java-squid/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValueRelation.java

```diff
@@ -28,20 +32,35 @@
 import java.util.List;
 import java.util.Set;
 
-public abstract class SymbolicValueRelation {
+public abstract class SymbolicValueRelation extends BinarySymbolicValue {
 
-  protected final SymbolicValue v1;
-  protected final SymbolicValue v2;
+  public SymbolicValueRelation(int id) {
+    super(id);
+  }
 
   protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
-    this.v1 = v1;
-    this.v2 = v2;
+    super(-1);
+    computedFrom(ImmutableList.of(v2, v1));
+  }
+
+  @Override
+  public BooleanConstraint shouldNotInverse() {
+    return BooleanConstraint.FALSE;
+  }
+
+  @Override
+  protected List<ProgramState> copyConstraint(SymbolicValue from, SymbolicValue to, ProgramState programState, BooleanConstraint booleanConstraint) {
+    ProgramState newState = programState;
+    if (programState.canReach(from) || programState.canReach(to)) {
+      newState = programState.addConstraint(this, booleanConstraint);
+    }
+    return super.copyConstraint(from, to, newState, booleanConstraint);
   }
 
   @Override
   public int hashCode() {
```

Comment:
Hashcode and equals relies on the class name which seems odd. I would add a Kind (as an int or an enum) for SVRelation (or rely on the one from syntax tree ? (not so good imo)) and use this kind for equals and hashcode. That will be far better perf wise and allow to not rely on reflection to make instanciation in tests but just on kind.